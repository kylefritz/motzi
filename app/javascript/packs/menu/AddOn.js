import React from "react";

import Quantity from "./Quantity";

export default class AddOn extends React.Component {
  state = { isChecked: false, quantity: 0 };
  constructor(props) {
    super(props);
    this.cbRef = React.createRef();
  }
  handleChanged() {
    const isChecked = this.cbRef.current.checked;
    this.setState({ isChecked });
    if (isChecked) {
      this.handleQuantityChanged(1);
    } else {
      // reset
      this.handleQuantityChanged(0);
    }
  }
  handleQuantityChanged(quantity) {
    this.setState({ quantity });
    this.props.onChange(quantity);
  }
  render() {
    const { name } = this.props;
    const { isChecked, quantity } = this.state;

    return (
      <div className="form-check mb-1">
        <label className="form-check-label">
          <input
            value={name}
            ref={this.cbRef}
            onChange={this.handleChanged.bind(this)}
            className="form-check-input"
            type="checkbox"
          />
          {isChecked && <strong className="pr-3">{quantity}x</strong>}
          {name}
        </label>
        {isChecked && (
          <div className="mb-4">
            <Quantity onChange={this.handleQuantityChanged.bind(this)} />
          </div>
        )}
      </div>
    );
  }
}
