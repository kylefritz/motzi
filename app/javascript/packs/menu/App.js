import React from 'react'
import axios from 'axios'
import * as Sentry from '@sentry/browser';
import queryString from 'query-string'

import Order from './Order.js'
import Menu from './Menu.js'

export default class App extends React.Component {
  componentDidMount() {
    const { uid } = queryString.parse(location.search)
    axios.get('/menu.json', { params: { uid } }).then(({ data }) => {
      this.setState(data) // expect: menu, user, order
      const { user } = data
      Sentry.configureScope((scope) => scope.setUser(user))
    }).catch((error) => {
      console.error("cant load menu", error)
      Sentry.captureException(error)
      this.setState({ error: "We can't load the menu" })
    })
  }

  handleCreateOrder(order) {
    if (this.state.order) {
      window.alert("Weird. This web site thinks you already placed an order. Refresh the page?!")
      return
    }

    console.debug('creating order', order)
    axios.post('/orders.json', order).then(({ data }) => {
      this.setState(data) // expect: menu, user, order
      console.debug('created order', data)
    }).catch((error) => {
      console.error("cant create order", error);
      window.alert("There was a problem submitting your order.")
      Sentry.captureException(error)
    });
  }

  render() {
    const { menu, user, order, error } = this.state || {};
    if (error) {
      return <>
        <h2 className="mt-5">{error}</h2>
        <p className="text-center">Sorry. Maybe try again or try back later?</p>
      </>
    }

    if (!menu) {
      return <h2 className="mt-5">loading...</h2>
    }

    if (order) {
      return <Order {...{ user, order, menu }} />
    }

    return <Menu {...{ user, order, menu, onCreateOrder: this.handleCreateOrder.bind(this) }} />
  }
}
