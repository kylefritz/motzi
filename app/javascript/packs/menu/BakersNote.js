import React from "react";
import { Converter } from "showdown";

export default function ({ bakersNote }) {
  const converter = new Converter({
    encodeEmails: false, // adds uncessary obfuscation to email addresses
    excludeTrailingPunctuationFromURLs: true,
    headerLevelStart: 3,
    simpleLineBreaks: true,
    simplifiedAutoLink: true,
  });
  const noteHtml = converter.makeHtml(bakersNote);
  return (
    <div
      className="bakers-note"
      dangerouslySetInnerHTML={{ __html: noteHtml }}
    />
  );
}
