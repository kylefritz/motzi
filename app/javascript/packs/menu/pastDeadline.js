import moment from "moment";

export function pastDeadline(deadline) {
  const now = moment();
  return now > moment(deadline);
}
