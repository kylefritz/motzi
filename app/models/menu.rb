class Menu < ApplicationRecord
    has_many :items, through: :menu_items
end
