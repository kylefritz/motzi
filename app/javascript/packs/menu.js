// Run by adding <%= javascript_pack_tag 'menu' %> to an erb page

import ErrorBoundary from "./ErrorBoundary";
import App from "./menu/App";

import React from "react";
import ReactDOM from "react-dom";

document.addEventListener("DOMContentLoaded", () => {
  ReactDOM.render(
    <ErrorBoundary>
      <App />
    </ErrorBoundary>,
    document.getElementById("react-menu")
  );
});
