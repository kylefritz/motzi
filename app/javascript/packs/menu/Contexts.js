import React, { useContext } from "react";
import { pastDeadline } from "./pastDeadline";
import _ from "lodash";

const DayContext = React.createContext();
const SettingsContext = React.createContext();

export { DayContext, SettingsContext };

export function getSettingsContext() {
  const ctx = useContext(SettingsContext);
  if (_.isNil(ctx)) {
    console.warn("SettingsContext is nil", ctx);
  }
  return ctx || {};
}

export function getDayContext() {
  const ctx = useContext(DayContext);
  if (ctx !== undefined) {
    const { day1Deadline, day2Deadline, ignoreDeadline } = ctx;

    const pastDay1Deadline = pastDeadline(day1Deadline);
    const day1Closed = pastDay1Deadline && !ignoreDeadline;
    const pastDay2Deadline = pastDeadline(day2Deadline);
    const day2Closed = pastDay2Deadline && !ignoreDeadline;

    return {
      ...ctx,
      pastDay1Deadline,
      pastDay2Deadline,
      day1Closed,
      day2Closed,
    };
  }

  return {
    day1: "Thursday",
    day1DeadlineDay: "Tuesday",
    pastDay1Deadline: false,
    day1Closed: false,
    //
    day2: "Saturday",
    day2DeadlineDay: "Thursday",
    pastDay2Deadline: false,
    day2Closed: false,
  };
}
