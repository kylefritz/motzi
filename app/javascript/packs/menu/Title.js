import React from "react";
import _ from "lodash";

import { getDayContext } from "./Contexts";

function When({ br }) {
  const { day1, day1DeadlineDay, day2, day2DeadlineDay } = getDayContext();

  return (
    <>
      Order by 9pm {day1DeadlineDay} for {day1} pickup {br && <br />}or 9pm{" "}
      {day2DeadlineDay} for {day2} pickup.
    </>
  );
}

export default function Title({ menu }) {
  const { name } = menu;
  const { pastDay2Deadline: isClosed } = getDayContext();

  return (
    <>
      <h2 id="menu-name">{name}</h2>
      {isClosed ? (
        <div
          className="alert alert-secondary text-center py-3 my-4"
          role="alert"
        >
          <h6 className="alert-heading">Ordering is closed for this menu</h6>
          <When menu={menu} br={true} />
        </div>
      ) : (
        <div id="deadline">
          <small>
            <When menu={menu} br={false} />
          </small>
        </div>
      )}
    </>
  );
}
