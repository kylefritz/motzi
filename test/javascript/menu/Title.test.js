import React from "react";
require("../configure_enzyme");
import { mount } from "enzyme";

import { DayContext } from "menu/Contexts";
import Title from "menu/Title";

//
// setup a mock for past deadline
//
import { pastDeadline } from "menu/pastDeadline";
jest.mock("menu/pastDeadline", () => {
  let areWePastDeadline = true;
  function __setPastDeadline(newValue) {
    areWePastDeadline = newValue;
  }

  return { __setPastDeadline, pastDeadline: () => areWePastDeadline };
});
import { __setPastDeadline } from "menu/pastDeadline";

test("__setPastDeadline back and forth in same test", () => {
  __setPastDeadline(false);
  expect(pastDeadline()).toBeFalsy();

  __setPastDeadline(true);
  expect(pastDeadline()).toBeTruthy();
});

function makeTitle() {
  return mount(
    <DayContext.Provider
      value={{
        day1: "Tuesday",
        day1DeadlineDay: "Sunday",
        day2: "Wednesday",
        day2DeadlineDay: "Monday",
      }}
    >
      <Title
        menu={{
          name: "Week 6: toast",
          orderingDeadlineText:
            "9:00 pm Tuesday for Thursday pickup or 9:00 pm Thurs for Sat pickup",
        }}
      />
    </DayContext.Provider>
  );
}
test("After deadline: ordering close", () => {
  __setPastDeadline(true);
  expect(makeTitle()).toMatchSnapshot();
});

test("Before deadline: small warning", () => {
  __setPastDeadline(false);
  expect(makeTitle()).toMatchSnapshot();
});
