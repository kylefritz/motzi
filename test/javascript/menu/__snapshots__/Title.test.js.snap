// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`After deadline: ordering close 1`] = `
<Title
  menu={
    Object {
      "name": "Week 6: toast",
    }
  }
>
  <h2
    id="menu-name"
  >
    Week 6: toast
  </h2>
  <div
    className="alert alert-secondary text-center py-3 my-4"
    role="alert"
  >
    <h6
      className="alert-heading"
    >
      Ordering is closed for this menu
    </h6>
    <When
      br={true}
      menu={
        Object {
          "name": "Week 6: toast",
        }
      }
    >
      Order by 9pm 
      Sunday
       for 
      Tuesday
       pickup 
      <br />
      or 9pm 
      Monday
       for 
      Wednesday
       pickup.
    </When>
  </div>
</Title>
`;

exports[`Before deadline: small warning 1`] = `
<Title
  menu={
    Object {
      "name": "Week 6: toast",
    }
  }
>
  <h2
    id="menu-name"
  >
    Week 6: toast
  </h2>
  <div
    id="deadline"
  >
    <small>
      <When
        br={false}
        menu={
          Object {
            "name": "Week 6: toast",
          }
        }
      >
        Order by 9pm 
        Sunday
         for 
        Tuesday
         pickup 
        or 9pm 
        Monday
         for 
        Wednesday
         pickup.
      </When>
    </small>
  </div>
</Title>
`;
